### ISSUE TITLE
Conan error when building Docker image.

### ISSUE THREAD

**Issue by wieringen** [valhalla/valhalla#3685]
This morning, I tried to build the Valhalla docker image, as I have been doing for the last couple of days. However, this time it failed to build. I tried again with a freshly cloned repo, but the issue remained:

```
#14 1.005 -- No compatible boost version detected, using conan...
#14 1.038 -- Conan: checking conan executable
#14 1.039 -- Conan: Found program /usr/local/bin/conan
#14 1.387 -- Conan: Version found Conan version 1.43.0
#14 1.387 -- Conan executing: /usr/local/bin/conan install /usr/local/src/valhalla/conanfile.txt --remote conancenter --settings build_type=RelWithDebInfo --settings compiler=gcc --settings compiler.version=9 --settings compiler.libcxx=libstdc++11 --settings compiler.cppstd=14
#14 2.116 WARN: Remotes registry file missing, creating default one in /root/.conan/remotes.json
#14 2.116 ERROR: boost/1.71.0: Cannot load recipe.
#14 2.116 Error loading conanfile at '/root/.conan/data/boost/1.71.0/_/_/export/conanfile.py': Unable to load conanfile in /root/.conan/data/boost/1.71.0/_/_/export/conanfile.py
#14 2.116   File "/usr/lib/python3.8/imp.py", line 171, in load_source
#14 2.116     module = _load(spec)
#14 2.116   File "<frozen importlib._bootstrap>", line 702, in _load
#14 2.116   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
#14 2.116   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
#14 2.116   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
#14 2.116   File "/root/.conan/data/boost/1.71.0/_/_/export/conanfile.py", line 1, in <module>
#14 2.116     from conan.tools.build import cross_building
#14 2.116 ImportError: cannot import name 'cross_building' from 'conan.tools.build' (/usr/local/lib/python3.8/dist-packages/conan/tools/build/__init__.py)
#14 2.116
#14 2.154 CMake Error at cmake/conan.cmake:638 (message):
#14 2.154   Conan install failed='1'
#14 2.154 Call Stack (most recent call first):
#14 2.154   CMakeLists.txt:153 (conan_cmake_install)
#14 2.154
#14 2.154
#14 2.154 -- Configuring incomplete, errors occurred!
#14 2.154 See also "/usr/local/src/valhalla/build/CMakeFiles/CMakeOutput.log".
#14 2.154 See also "/usr/local/src/valhalla/build/CMakeFiles/CMakeError.log".
------
executor failed running [/bin/sh -c cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_C_COMPILER=gcc]: exit code: 1
```

Any idea what the problem might be?

**Comment by nilsnolde** [comment]
argh, that's annoying.. I figure what happens is that you have cached layers of the valhalla image locally which has a fairly old conan version (1.43.0 as per logs). the layer is this line: https://github.com/valhalla/valhalla/blob/master/docker/Dockerfile-build#L4. I did a bit of digging: the import the conan boost recipe is failing on [wasn't added until conan v1.43.4 or so](https://github.com/conan-io/conan/pull/10536/files#diff-74272a6f8877a3af1e97d287b5e9169050b4c836c436d7d3779cb6842dacef65), definitely not there yet in 1.43.0.

it looks like the boost recipe for conan [introduced a bug roughly 24 hours ago](https://github.com/conan-io/conan-center-index/commit/abf40e34637400c5c3f625d230098bfaee39bc06#diff-cc14160728d1ad321c1444ffad82615139181e681f4acb7dfd045a005cc8bd66), which would fit with your experience. you're really unlucky :sweat_smile: it's pretty exotic I think, I still filed [an issue](https://github.com/conan-io/conan-center-index/issues/11955).

there's a few workarounds coming to mind, but some would involve adding pretty hacky docker instructions. I think the best in these cases is you disable layer caching while building (`--no-cache` flag). you only have to do that once, so you cache a new layer with a fresher conan version. still, this might happen more often in the future, the file with the bug is shared by all python & (almost) all conan versions, seems fairly vulnerable for regressions.

I'll close, feel free to re-open if that didn't work.

**Comment by nilsnolde** [comment]
actually, I should leave this open. anyone who installed conan a while ago, e.g. for dev, will run into the same problem. in that case it's easy though, `pip install --upgrade conan`

**Comment by ImreSamu** [comment]
@nilsnolde 

imho: this issue is related to `/docker/Dockerfile-run`
and your suggestions ( `pip install --upgrade conan` ) is working.  I have created a PR  https://github.com/valhalla/valhalla/pull/3689

Not urgent, but now the [builder image `FROM valhalla/valhalla:build-3.1.5 as builder` ](https://github.com/valhalla/valhalla/blob/master/docker/Dockerfile-run#L3) older than 7 month ( Created: `2021-12-14` ) 
and should be updated so 
not only the `conan` library is updated, but also `gcc` from 9.3 to 9.4.  ( [more than 190 bugs fixed ! ](https://gcc.gnu.org/pipermail/gcc-announce/2021/000168.html)) 

```
$ docker inspect --format='{{index .RepoDigests 0}} {{.Created}}' valhalla/valhalla:build-3.1.5 
valhalla/valhalla@sha256:b5d2e91c90e00685b25e0396d945789b72126cc75300a832c5ed5ad0ad7bbffd 2021-12-14T16:38:19.220448588Z
```

**Comment by nilsnolde** [comment]
thanks @ImreSamu ! 

IMHO we should move to valhalla 3.2.0 actually :) there's a few PRs I guess we'd like to include (notable removing geos in #3683) , but then again there are always a few PRs we'd like to wait for haha should we just finally release after that one PR @kevinkreiser ?

I'm definitely not wondering about this the first time and I sure miss smth: why exactly are we referencing the `3.1.5` image in the `master` branch? with circle ci we build and publish from master as `build-latest`, so the `run-latest` image that's being built from master should actually take `build-3.1.5` image as base rather than `build-latest` no? but that can't be, otherwise the `run-latest' image would have pretty ancient code.. can someone jog my memory? some line I'm missing somewhere?

**Comment by kevinkreiser** [comment]
The reason it uses that is because that is the last time we changed the base build image which doesn't have our code in it just the tools needed to build our code. Remember that all the builds need thus base build image to run. If you change something in there we would have to have the builds wait for it to land on docker hub before they start. Instead of that we just pin to the last changes we versioned.  I'm not even sure how we could do that anyway in ci. I guess using latest might give the false idea that you are actually using latest when really you would be using the previous latest haha. Then again using tags for the builder container that look like the software version tags makes it confusing in the way that you are pointing out. We could just change the builder version to have its own versioning

**Comment by nilsnolde** [comment]
> If you change something in there we would have to have the builds wait for it to land on docker hub before they start.

but ci builds the `build-` image before the `run-` image anyways. so in that ci run, we can just use what docker already knows about (`build-latest` as base image for the `run-latest` image), no need to pull smth from dockerhub (or wait for smth to be uploaded), it seems to me at least. it would in fact be faster than right now, since it's pulling the `build-3.1.5` from dockerhub on every run. the ci yaml wouldn't have to change. we could introduce a docker build arg in the `FROM` clause and switch between tagged and latest, less release hassle. I'll try to draw smth up quickly. 

I don't see the hole in that logic, but doesn't mean anything :sweat_smile: does that make sense to you @kevinkreiser ?

> I guess using latest might give the false idea that you are actually using latest when really you would be using the previous latest haha

but that's what it is right now, isn't it? you pull `run-latest` and get it based off a `build-dec2022` image, where I'd expect a base image/operating system that's equally "latest". in fact it looks like that's the reason that this conan bug was triggered. we released the 3.1.5 image in december 2022. that's [the date of the conan version](https://github.com/conan-io/conan/releases/tag/1.43.0) pretty much. also that's a pretty old version of ubuntu missing security patches etc. not that super relevant if one only uses valhalla in a container, but who knows. it doesn't bite us in ci bcs we never again actually build the tagged image, we just pull the dec 2022 image from dockerhub.

**Comment by nilsnolde** [comment]
ah sorry, I know now what you mean, the circle ci workflow that runs in parallel.. right. hmpf

**Comment by nilsnolde** [comment]
hmm.. I know a full docker build takes forever, around 50 mins apparently. but how about paying that penalty on master pushes but not on PR's (currently we're also not doing that on PRs except for building the `build` image but then not even use it)? would that be prohibitively expensive? in more recent past we had around max 4 master pushes per week (not counting fairly rare direct master pushes, only merges and on average more around 2-3 per week). so we'd spend max 16 hours CI hours more per month. that feels at least dwarfed by all the PR pushes where each ci run takes also around one hour and it's surely more than 5 times as many PR runs as there's merges. there's people pushing every typo causing ci runs (exaggerated, but it does kinda bug me, it's lots of energy wasted). since it's already approved I wouldn't be too bothered by the fact that it takes almost 2 hours for CI on master to be finished. it could be annoying when one is in a merge frenzy :smile: I guess that does happen now and then..

anyways, to me it'd be worth it to spend the extra 5-20% CI hours, so we can have fresh a base image on the `run-latest`. happy to pitch in if that's a concern and it's possible. what do you think?

**Comment by kevinkreiser** [comment]
we could do that on master but then there are 2 problems:

1. how do we make the other builds wait for that workflow to finish
2. masters build outcome could be totally different than the pr that just merged because it now uses a different build image

yeah im still trying to understand the primary benefit of being able to target the latest build image. i feel like i dont understand what we are missing out on when the build image changes so infrequently. just to be clear, when we do need to add or change how the build image works we basically have to go through the pain of making those changes first in a separate pr, getting that merged and then a follow up pr that takes advantage of whatever change was there. i totally agree that that is annoying but its very infrequent, like way more infreqent then merging master. anyway i feel like you have a specific thing that you are running into with this so where this is more painful than making 2 prs, bring me up to speed :D

one thing that comes to mind to me though is that all of this wouldnt matter if we could just do the thing that we do with the mac build, spend a minute installing the dependencies that it needs and never worry about starting with build container. so the question is how long does it take to run that dependency install script? or better yet, can we get CI to cache the result of installing the crap in that script? in other words if we could get the time down that it takes to get the deps we need to build we wouldnt even need a docker container with the build stuff in it. but maybe that is too crazy an idea?

**Comment by nilsnolde** [comment]
good points. I think I just don't really like that all software versions & os patches are a bit old (when I think about my os status), sometimes triggering some bugs like this one (though we can't protect yourself against all bugs in other projects), while suggesting with the image's name I'd actually have a "latest" base system. that's my main point I guess. I do realize I dedicated more words than it really deserves though :sweat_smile: 

> how long does it take to run that dependency install script

that's true, it's very quick actually.. only takes [2 minutes](https://app.circleci.com/pipelines/github/valhalla/valhalla/9307/workflows/69399cb5-e8c6-4328-bbb8-6358f5fcfc9f/jobs/53583), so the docker instructions could run inside of each job first thing and moved to the Dockerfile-run. and remove it? 

> 1. how do we make the other builds wait for that workflow to finish

cicle ci also supports a `requires` list of other jobs to finish before a job is started. we could put the docker workflow in the `requires` for all other jobs and only actually build anything docker when it's master or tagged, so no delay for PRs.

> 2. masters build outcome could be totally different than the pr that just merged because it now uses a different build image

true, but that's the same right now, no? we don't build the `run-` image of a PR before we don't merge it in master. although we do run the other PR jobs on base of the new `build-` image after changes were merged to master. we would be somewhere similar if we'd instead install all deps on each run fresh. then other workflow's base image could be ubuntu:20.04 which would exactly reproduce what's happening inside the docker image. and in that case we wouldn't even need to schedule workflows sequentially with `requires`.

am I overly concerned about details? :sweat_smile:

**Comment by nilsnolde** [comment]
well, actually it's a multi-stage build for the `run-` image, so the final image is a pretty up-to-date ubuntu system, just valhalla is built within an older one, who cares. so many interdependent layers & processes.. so I guess, never mind then when I almost undermine my only half-esoteric reasoning? :sweat_smile:

**Comment by nilsnolde** [comment]
> I'm definitely not wondering about this the first time and I sure miss smth:

not kidding, we definitely had that discussion before and I'm a bit afraid in 2 years I'll be like "why are we actually basing the run- image of a tagged image?" hahaha sorry for spam

**Comment by kevinkreiser** [comment]
so if it only takes 2 minutes, maybe we can indeed just switch back to running the script and getting rid of the build container. i love simplifying stuff so i'm totally down with that. is it an interesting approach to you @nilsnolde ?

**Comment by wieringen** [comment]
To speed up docker CI builds for my own projects, I have been using [Kaniko](https://github.com/GoogleContainerTools/kaniko).

Kaniko can cache layers created by RUN and COPY commands in a remote repository. Before executing a command, Kanika checks the cache for the layer. If it exists, kaniko will pull and extract the cached layer instead of executing the command. If not, kaniko will execute the command and then push the newly created layer to the cache.

Works pretty well. It replaces the docker build and push command.

**Comment by nilsnolde** [comment]
thanks @wieringen! did you ever experiment with `--build-arg BUILDKIT_INLINE_CACHE=1` for buildkit docker builds? here it sounds like it's able to cache to/from the remote docker registry: https://testdriven.io/blog/faster-ci-builds-with-docker-cache/#buildkit. that'd be really nice too:)
